File: src/main/java/ru/ioleksiv/telegram/bot/core/controller/handler/check/impl/BooleanValidator.java
package ru.ioleksiv.telegram.bot.core.controller.handler.check.impl;

import ru.ioleksiv.telegram.bot.api.model.annotation.BooleanState;
import ru.ioleksiv.telegram.bot.core.controller.handler.check.Validator;

public class BooleanValidator implements Validator<Boolean> {

    private final BooleanState booleanState;

    public BooleanValidator(BooleanState booleanState) {
        this.booleanState = booleanState;
    }

    @Override
    public boolean check(Boolean argument) {

        if (booleanState == BooleanState.NON_NULL) {
            return true;
        }

        if (booleanState == BooleanState.TRUE) {
            return argument;
        }

        if (booleanState == BooleanState.FALSE) {
            return !argument;
        }

        throw new UnsupportedOperationException("Unexpected boolean state");
    }

}

File: core/src/main/kotlin/io/github/oleksivio/tl/kbot/core/core/controller/handler/check/impl/BooleanValidator.kt
package io.github.oleksivio.tl.kbot.core.core.controller.handler.check.impl

import io.github.oleksivio.tl.kbot.core.core.controller.handler.check.Validator
import io.github.oleksivio.tl.kbot.core.model.annotation.BooleanState

class BooleanValidator(private val booleanState: BooleanState) :
    Validator<Boolean> {

    override fun invoke(checkedValue: Boolean): Boolean {

        if (booleanState === BooleanState.NON_NULL) {
            return true
        }

        if (booleanState === BooleanState.TRUE) {
            return checkedValue
        }

        if (booleanState === BooleanState.FALSE) {
            return (!checkedValue)
        }

        throw UnsupportedOperationException("Unexpected boolean state")
    }
}
